File: Lectures/Lecture 3/python/queue.py
from linked_list import LinkedList


class Queue(LinkedList):
    def enqueue(self, item):
        """ Add item to start of the linked list """
        self.prepend(item)

    def dequeue(self):
        """ Remove and return item from end of the linked list """
        data = self.tail.data
        self.remove_last_node()
        return data

    def is_empty(self):
        return self.head is None

    def peek(self):
        if self.tail is None:
            return None

        return self.tail.data

    def size(self):
        pass


if __name__ == '__main__':
    my_queue = Queue()
    print("my_queue:", my_queue)

    print("Empty?", my_queue.is_empty())
    print()

    my_queue.enqueue("hello")
    print("my_queue:", my_queue)

    my_queue.enqueue("world")
    print("my_queue:", my_queue)

    print("Empty?", my_queue.is_empty())
    print("peek:", my_queue.peek())
    print()

    print("dequeue:", my_queue.dequeue())
    print("my_queue:", my_queue)
    print("dequeue:", my_queue.dequeue())
    print("my_queue:", my_queue)

    print("peek:", my_queue.peek())

File: Lectures/Lecture 3/cpp/queue.cpp
#include "linked_list.cpp"

using namespace std;

class Queue : public LinkedList {
public:
    void enqueue(int data) {
        this->prepend(data);
    }

    int dequeue() {
        int data = this->tail->data;
        this->remove_last_node();
        return data;
    }

    bool is_empty() {
        return (this->head==NULL);
    }

    int peek() {
        if (this->tail == NULL) {
            throw std::out_of_range("Queue is empty");
        }

        return this->tail->data;
    }

    // Size is defined in base class, no need to redefine
};

main() {
    Queue my_queue = Queue();
    cout << "My queue: " << my_queue << endl;
    
    cout << boolalpha;   // Ensure we can print booleans as true/false instead of 1/0
    cout << "Empty?: " << my_queue.is_empty() << endl;
    cout << endl;
    
    my_queue.enqueue(1);
    cout << "My queue: " << my_queue << endl;
    my_queue.enqueue(2);
    cout << "My queue: " << my_queue << endl;

    cout << "Empty?: " << my_queue.is_empty() << endl;
    cout << "Size: " << my_queue.size() << endl;
    cout << "Peek: " << my_queue.peek() << endl;
    cout << endl;

    cout << "Dequeue: " << my_queue.dequeue() << endl;
    cout << "My queue: " << my_queue << endl;
    cout << "Dequeue: " << my_queue.dequeue() << endl;
    cout << "My queue: " << my_queue << endl;

     cout << "Peek: " << my_queue.peek() << endl;   
}